### PR TITLE
Sync Batch Norm for PyTorch

### LINKED ISSUE #1384: Sync Batch Norm for Pytorch
There are sync batch norm implementations for PyTorch distributed and Apex but can't seem to find one for Horovod. Is there a technical reason why there isn't one?

### PR DESCRIPTION
**Pull request by alsrgv**
Implementation of https://pytorch.org/docs/stable/nn.html#syncbatchnorm using Horovod.  Current version uses optimized CUDA kernels written for PyTorch and have two different invocations because they changed.  We can consider making our own implementation if it turns out to be a hassle.

## Why Sync Batch Norm?
As evidenced by https://arxiv.org/abs/1804.07612, small batches are great for training neural networks.  However, https://arxiv.org/abs/1803.08494 noted that multi-GPU training with small-batch BatchNorm is detrimental to performance.

SyncBatchNorm improves training where each worker can only hold few examples (1..4) and total number of workers is not too high - at which point it starts to lose its regularization abilities.

Fixes #1384

### CODE REVIEW

File: horovod/torch/sync_batch_norm.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        count_handle = allgather_async(count.unsqueeze(0), name='sync_batch_norm.count')
+        mean_handle = allgather_async(mean.unsqueeze(0), name='sync_batch_norm.mean')
+        invstd_handle = allgather_async(invstd.unsqueeze(0), name='sync_batch_norm.invstd')
+
+        # wait on the async communication to finish
+        count_all = synchronize(count_handle)
+        mean_all = synchronize(mean_handle)
+        invstd_all = synchronize(invstd_handle)
+
+        if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
```
Comment (tgaddair) [inline horovod/torch/sync_batch_norm.py:109]: Loose version check are somewhat expensive.  Maybe move this to a constant?

File: horovod/torch/sync_batch_norm.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        self.save_for_backward(input, weight, mean, invstd, count_all)
+
+        # apply element-wise normalization
+        return torch.batch_norm_elemt(input, weight, bias, mean, invstd, eps)
+
+    @staticmethod
+    def backward(self, grad_output):
+        grad_output = grad_output.contiguous()
+        saved_input, weight, mean, invstd, count_all = self.saved_tensors
+        grad_input = grad_weight = grad_bias = None
```
Comment (tgaddair) [inline horovod/torch/sync_batch_norm.py:136]: Seems `grad_weight` and `grad_bias` are always set.

File: horovod/torch/sync_batch_norm.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+        # calculate local stats as well as grad_weight / grad_bias
+        sum_dy, sum_dy_xmu, grad_weight, grad_bias = torch.batch_norm_backward_reduce(
+            grad_output,
+            saved_input,
+            mean,
+            invstd,
+            weight,
+            self.needs_input_grad[0],
```
Comment (tgaddair) [inline horovod/torch/sync_batch_norm.py:145]: Can we explode the `self.needs_input_grad` tuple to something like:

```
need_input_grad, need_weight_grad, need_bias_grad, _, _, _ = self.needs_input_grad
```

for clarity?

File: horovod/torch/sync_batch_norm.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        if self.needs_input_grad[0]:
+            # synchronizing stats used to calculate input gradient.
+            sum_dy_handle = allreduce_async(sum_dy, op=Sum, name='sync_batch_norm.sum_dy')
+            sum_dy_xmu_handle = allreduce_async(sum_dy_xmu, op=Sum, name='sync_batch_norm.sum_dy_xmu')
+
+            # wait on the async communication to finish
+            sum_dy = synchronize(sum_dy_handle)
+            sum_dy_xmu = synchronize(sum_dy_xmu_handle)
+
+            if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
```
Comment (tgaddair) [inline horovod/torch/sync_batch_norm.py:159]: Same as above, consider moving this to a constant.

File: horovod/torch/sync_batch_norm.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    @torch.jit.unused
+    def _maybe_run_sync_bn(self, input):
+        if size() == 1:
+            return self._run_bn(input)
+        return _SyncBatchNorm.apply(
+            input, self.weight, self.bias, self.running_mean, self.running_var,
+            self.eps, self.momentum)
+
+    def forward(self, input):
+        # currently only GPU input is supported
```
Comment (tgaddair) [inline horovod/torch/sync_batch_norm.py:74]: Can you add a TODO stating what would be needed to get it work on CPU?

File: horovod/torch/sync_batch_norm.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+if not hasattr(torch.jit, 'unused'):
+    torch.jit.unused = lambda x: x
+
+
+class SyncBatchNorm(_BatchNorm):
+    """
+    Applies synchronous version of N-dimensional BatchNorm.  In this version, normalization
+    parameters are synchronized across workers during forward pass.  This is very useful in
+    situations where each GPU can fit a very small number of examples.
+
```
Comment (tgaddair) [inline horovod/torch/sync_batch_norm.py:37]: Can you include a comment in the docstring about requiring the tensor to be on GPU?

File: test/test_torch.py [hunk 7]
```diff
@@ -1667,6 +1667,59 @@ def test_horovod_join_broadcast(self):
                 ret = hvd.join(hvd.local_rank())
             else:
                 ret = hvd.join()
+    
+    def test_horovod_sync_batch_norm(self):
```
Comment (tgaddair) [inline test/test_torch.py:1671]: May also want to include a skip if Python version is < 3 (integration tests passing because we have no Python 2 GPU tests anymore).

File: horovod/torch/sync_batch_norm.py (hunk 5 above)
Comment (alsrgv) [reply horovod/torch/sync_batch_norm.py:74]: This is constrained by PyTorch internal kernels that we use. They're only implemented on GPU.  If we want to make it work on CPU, we should consider rolling our own implementation w/o dependency on PyTorch internals.

File: horovod/torch/sync_batch_norm.py (hunk 1 above)
Comment (alsrgv) [reply horovod/torch/sync_batch_norm.py:109]: Fixed

File: horovod/torch/sync_batch_norm.py (hunk 2 above)
Comment (alsrgv) [reply horovod/torch/sync_batch_norm.py:136]: Fixed

File: horovod/torch/sync_batch_norm.py (hunk 3 above)
Comment (alsrgv) [reply horovod/torch/sync_batch_norm.py:145]: Fixed

File: horovod/torch/sync_batch_norm.py (hunk 4 above)
Comment (alsrgv) [reply horovod/torch/sync_batch_norm.py:159]: Fixed

File: horovod/torch/sync_batch_norm.py (hunk 6 above)
Comment (alsrgv) [reply horovod/torch/sync_batch_norm.py:37]: Done

File: test/test_torch.py (hunk 7 above)
Comment (alsrgv) [reply test/test_torch.py:1671]: Added skip, but we should deprecate all Py2 tests :-)

Comment (tgaddair) [reply test/test_torch.py:1672]: Agreed.  Tentative plan is to drop Python 2 in 0.20.0.

File: horovod/torch/sync_batch_norm.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        size = input.numel() // input.size(1)
+        count = torch.tensor([size])
+
+        # calculate mean/invstd for input.
+        mean, invstd = torch.batch_norm_stats(input, eps)
+
+        count_handle = allgather_async(count.unsqueeze(0), name='sync_batch_norm.count')
+        mean_handle = allgather_async(mean.unsqueeze(0), name='sync_batch_norm.mean')
+        invstd_handle = allgather_async(invstd.unsqueeze(0), name='sync_batch_norm.invstd')
```
Comment (thuyen) [inline horovod/torch/sync_batch_norm.py:107]: We would have a lot of `BatchNorm` instances in a network. Would we need to add the instance `id` to the `name` arguments in `allreduce_async` calls here and below (line 157, 158) ?

Comment (alsrgv) [reply horovod/torch/sync_batch_norm.py:107]: Didn't seem necessary so far - current implementation of SyncBN requires each worker to execute model in the same sequence with respect to batch norms.

Comment (thuyen) [reply horovod/torch/sync_batch_norm.py:107]: Got you thanks!

File: horovod/torch/sync_batch_norm.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+            sum_dy_xmu = synchronize(sum_dy_xmu_handle)
+
+            if _SYNC_BN_V2:
+                mean_dy = sum_dy / count_all.sum()
+                mean_dy_xmu = sum_dy_xmu / count_all.sum()
+            else:
+                # before 1.6.0, sum_dy was sum of means from every worker, so we just 
+                # need to divide it by number of workers
+                mean_dy = sum_dy / size()
+                mean_dy_xmu = sum_dy_xmu / size()
```
Comment (thuyen) [inline horovod/torch/sync_batch_norm.py:171]: I think we should divide by `count_all.sum()` for both pytorch 1.5 and 1.6.
 
For pytorch 1.5, right now if I comment out `_run_bn` code path (not using torch.nn.BatchNorm) and run the following on `v0.19.3` locally:
```
     torch.cuda.manual_seed(2020)
     bn = torch.nn.BatchNorm2d(10).cuda()
     bn_hvd = SyncBatchNorm(10).cuda()

     x = torch.rand(3, 10, 8, 8).cuda()
     x1 = x.clone().requires_grad_()
     x2 = x.clone().requires_grad_()
     y = bn(x1)
     y_hvd = bn_hvd(x2)

     y.sum().backward()
     y_hvd.sum().backward()
     print((x1.grad - x2.grad).abs().sum())
```
I got 
```
tensor(1298650., device='cuda:0')
```
But if  if we divide by `count_all.sum()` here, I got:
```
tensor(0.0004, device='cuda:0')
```

Comment (alsrgv) [reply horovod/torch/sync_batch_norm.py:171]: Great catch, I've raised #1980 to address this.  Apparently they made breaking changes in both 1.5.0 and 1.6.0.

**Comment by eric-haibin-lin** [comment]
Does SyncBatchNorm support the use case where I have 2 nodes, each with 8 GPUs, and the sync only happens intra-node instead of inter-node (sync 2 groups of 8 workers instead of 16)

**Comment by alsrgv** [comment]
@eric-haibin-lin, this implementation does not, but please feel free to extend it!

**Comment by romerojosh** [comment]
@alsrgv @eric-haibin-lin I think the main limitation here is that Horovod still is currently designed around performing collectives across the global communicator, with no options currently to perform collectives on subsets of workers. @tgaddair, perhaps this is something we should begin to think about.

**Comment by tgaddair** [comment]
Good point, @romerojosh.  Added #2139 to track.
